extras/dump_iphone_video.py

```diff
@@ -1,39 +1,167 @@
+    # Connect to the iPhone
+    logger.debug('Connecting to iPhone')
+    phone = iphone.IPhone("dump_iphone")
+    handshake_success = phone.prepare()
+    if not handshake_success:
+        logger.error(f'Unable to connect to iPhone [state={phone._state}]!')
+        return
 
-    iphone.disconnect()
+    # Get a list of stored files
+    flist = phone.dumpall_getfilelist(log_files=False)
+    if flist is None:
+        logger.error(f'Unable to retrieve file list [state={phone._state}]!')
+        phone.disconnect()
+        return
+    logger.debug(f'{len(flist)} files to transfer: {str(flist)}')
 
+    # Try to extract and save each file
+    for fname in flist:
+        # Parse the session folder out of the file name
+        sess_name = re.findall("[0-9]*_[0-9]{4}-[0-9]{2}-[0-9]{2}", fname)
+        if len(sess_name) == 0 or sess_name is None:
+            logger.error(f'Invalid session name: file={fname}; name={sess_name}.')
+            continue
 
-if __name__ == "__main__":
-    import datetime
+        # Make the session folder if it does not exist
+        sess_folder = op.join(session_root, sess_name[0])
+        if not op.exists(sess_folder):
+            logger.debug(f'Creating directory: {sess_folder}')
+            os.mkdir(sess_folder)
+
+        try:
+            dump_file(
+                phone, fname, op.join(sess_folder, fname),
+                timeout_sec=args.timeout,
+                delete_zero_byte=args.delete_zero_byte,
+            )
+        except TimeoutException:
+            logger.warning('Discontinuing file transfer to prevent out-of-order files.')
+            break
+
+    logger.debug('Disconnecting iPhone')
+    phone.disconnect()
+
+
+def dump_file(
+        phone: iphone.IPhone,
+        fname: str,
+        fname_out: str,
+        timeout_sec: Optional[float] = None,
+        delete_zero_byte: bool = False,
+) -> None:
+    """
+    Extract a single file from the iPhone, save it in the specified location, and tell the iPhone to delete the file.
+
+    Parameters
+    ----------
+    phone
+        The iPhone object to interface with.
+    fname
+        The name of the file on the iPhone (returned by dumpall_getfilelist).
+    fname_out
+        The path to save the retrieved file to.
+    timeout_sec
+        If not None, log an error and return if the file transfer exceeds the timeout.
+    delete_zero_byte
+        If true, delete files from the iPhone if no data is observed.
+    """
+    logger = logging.getLogger('iphone_dump')
+    if op.exists(fname_out):  # Do not overwrite a file that already exists
+        logger.error(f'Cannot write {fname_out} as it already exists!')
+        return
+
+    # Attempt to retrieve the file from the iPhone
+    logger.info(f'Dump {fname} -> {fname_out}')
+    success, file_data = phone.dump(fname, timeout_sec=timeout_sec)
+    if not success:
+        logger.error(f'Unable to retrieve {fname}! (E.g., wait timed out.)')
+        raise TimeoutException
```

Comment:
I could get behind this solution, but then for uniformity I think every call that sends a message should raise a TimeoutException were appropriate. This is somewhat problematic as the iphone currently swallows most of these to avoid crashing the server script... we could maybe set an __init__ argument that toggles whether they get swallowed.

Reply:
I see. For me, the real issue here is that the client code gets False, and then says, oh, I only get False if there was a timeout, so I'll throw a TimeoutException. This may be true today, but the iPhone code could easily be modified in the future to return False for some other reason and then the wrong exception gets thrown.  

What about instead of returning False create a string enumeration that returns "TIMEOUT" so it could be extended in the future to handle other failure modes?

Reply:
Yea, I see two options. Both rework all timeout-related function calls for consistency:
1. Have a result enum that for now is SUCCESS and TIMEOUT. Rework all relevant calls to return this result flag.
2. Have a raise_timeout() call that only raises the exception if specified by the constructor. Rework all relevant calls to use raise_timeout and get rid of the success flag on dump.

Reply:
Currently leaning towards the latter (and then we just enable the timeout exception for the dump script and catch/handle it there), but what do you think?

The raise_timeout function will be something like:
```
def raise_timeout(self):
    self.logger.error('Timeout encountered when waiting for response.')
    if self.enable_timeout_exceptions:
        raise IPhoneTimeout()
```

Reply:
Went ahead with approach 2, I think it's a bit cleaner and required the least changes across the code.